frontend/catalyst/jax_extras/patches.py

```diff
@@ -180,3 +183,13 @@ def _gather_shape_rule_dynamic(
             )
 
     return _gather_shape_computation(indices, dimension_numbers, slice_sizes)
+
+
+def _sin_lowering2(ctx, x):
+    """Use hlo.sine lowering instead of the new sin lowering from jax 0.4.28"""
+    return _nary_lower_hlo(hlo.sine, ctx, x)
+
+
+def _cos_lowering2(ctx, x):
+    """Use hlo.cosine lowering instead of the new cosine lowering from jax 0.4.28"""
+    return _nary_lower_hlo(hlo.cosine, ctx, x)
```

Comment:
Just out of curiosity, do you have some more details here, why is the new lowering incompatible? :)

Reply:
The new version triggers this error. 
`TypeError: Shapes must be 1D sequences of integer scalars, got (Var(id=137689073075072):int32[], Var(id=137689073075136):int32[])`

`mlir.lower_fun` is the new thing as follows. It tries to get shape but the shape is in tuple in list. 
`def _sin_lowering(ctx, x):
  if dtypes.issubdtype(ctx.avals_in[0].dtype, np.complexfloating):
    sine = mlir.lower_fun(_sin_complex, multiple_results=False)
    return sine(ctx, x)
  return _nary_lower_hlo(hlo.sine, ctx, x)`
https://github.com/google/jax/blob/ccc4c42ec9d0625fd25a7361acf241f41fd2109e/jax/_src/lax/lax.py#L1928-L1932

Reply:
The trace:
```python
frontend/test/pytest/test_jax_dynamic_api.py:139 (test_classical_tracing_unary_ops[sin])
>   assert_array_and_dtype_equal(f(shape), op(jnp.ones(shape, dtype)))

test_jax_dynamic_api.py:157: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../../../.conda/envs/xanadu-update/lib/python3.12/site-packages/pennylane/logging/decorators.py:61: in wrapper_entry
    return func(*args, **kwargs)
../../catalyst/jit.py:454: in __call__
    requires_promotion = self.jit_compile(args)
../../../../../.conda/envs/xanadu-update/lib/python3.12/site-packages/pennylane/logging/decorators.py:61: in wrapper_entry
    return func(*args, **kwargs)
../../catalyst/jit.py:525: in jit_compile
    self.jaxpr, self.out_type, self.out_treedef, self.c_sig = self.capture(args)
../../catalyst/debug/instruments.py:143: in wrapper
    return fn(*args, **kwargs)
../../../../../.conda/envs/xanadu-update/lib/python3.12/site-packages/pennylane/logging/decorators.py:61: in wrapper_entry
    return func(*args, **kwargs)
../../catalyst/jit.py:587: in capture
    jaxpr, out_type, treedef = trace_to_jaxpr(
../../../../../.conda/envs/xanadu-update/lib/python3.12/site-packages/pennylane/logging/decorators.py:61: in wrapper_entry
    return func(*args, **kwargs)
../../catalyst/jax_tracer.py:531: in trace_to_jaxpr
    jaxpr, out_type, out_treedef = make_jaxpr2(func, **make_jaxpr_kwargs)(*args, **kwargs)
../../catalyst/jax_extras/tracing.py:555: in make_jaxpr_f
    jaxpr, out_type, consts = trace_to_jaxpr_dynamic2(f)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   return op(jnp.ones(s, dtype))
E   jax._src.source_info_util.JaxStackTraceBeforeTransformation: TypeError: Shapes must be 1D sequences of integer scalars, got (Var(id=133767151424768):int32[], Var(id=133767151424832):int32[])
E   
E   The preceding stack trace is the source of the JAX operation that, once transformed by JAX, triggered the following exception.
E   
E   --------------------
```

Reply:
Can you add a test here where sine and cosine are used with a complex value as an operand?

Reply:
Added. I will also open an issue for this sine/cosine lowering fix.

Reply:
I think the [test is already working with complex numbers](https://github.com/PennyLaneAI/catalyst/blob/5fa4b21922ab1e7beb8f83cd1a8daf4b0c298c95/frontend/test/pytest/test_jax_dynamic_api.py#L150-L157), the error looks to me like a dynamic shape issue. I guess this new sine lowering doesn't work with it while the old one did 🤔

Reply:
In the interest of completing the upgrade I would be okay merging the patch for now, unless we have any ideas what else to do. @tzunghanjuang Could you create an issue in our repo regarding this?

Reply:
I have created https://github.com/PennyLaneAI/catalyst/issues/972.
I thought Erick means an separate test for scalar complex value. I have removed that and add cosine to the existing one.

Reply:
Thanks!